### PR TITLE
Updating snapshots of a TaskRunner when status of Task changes

### PR DESCRIPTION
**Pull request by diptanu**
This PR updates the snapshot only for a TaskRunner when it's state is updated. This doesn't fix the client errors while restarting if the drivers can't attached to running or failed tasks.

### CODE REVIEW

File: client/alloc_runner.go [hunk 1]
```diff
@@ -257,6 +269,7 @@ func (r *AllocRunner) setTaskStatus(taskName, status, desc string) {
 		Description: desc,
 	}
 	r.taskStatusLock.Unlock()
+	r.SaveState(taskName)
```
Comment (dadgar) [inline client/alloc_runner.go]: Just get the task runner here and call `tr.SaveState()` on it.

File: client/alloc_runner.go [hunk 2]
```diff
@@ -119,7 +119,7 @@ func (r *AllocRunner) RestoreState() error {
 }
 
 // SaveState is used to snapshot our state
-func (r *AllocRunner) SaveState() error {
+func (r *AllocRunner) SaveState(taskName string) error {
```
Comment (dadgar) [inline client/alloc_runner.go]: I would revert this change. You are trying to do two things in one method.

File: client/alloc_runner.go (hunk 1 above)
Comment (diptanu) [inline client/alloc_runner.go]: @dadgar We are also logging and adding the error to multierror, and it was being repeated in two places so it made sense to extract that as a method.

File: client/alloc_runner.go (hunk 2 above)
Comment (diptanu) [inline client/alloc_runner.go]: Yeah makes sense, we can do better.

File: client/alloc_runner.go (hunk 1 above)
Comment (dadgar) [inline client/alloc_runner.go]: Sure you can put it in a method that takes task name and returns an error but you shouldn't be passing the multi error into it. You are also eating the error here so not sure it matters.

Comment (diptanu) [inline client/alloc_runner.go]: It's not enough to save the Task Runner state, since the Alloc state could have changed too, once the Task state has changed

File: client/alloc_runner.go [hunk 3]
```diff
@@ -128,23 +147,19 @@ func (r *AllocRunner) SaveState() error {
 		Context:       r.ctx,
 	}
 	err := persistState(r.stateFilePath(), &snap)
-	r.taskStatusLock.RUnlock()
 	if err != nil {
 		return err
 	}
+	return nil
```
Comment (dadgar) [inline client/alloc_runner.go]: `return persistState(r.stateFilePath(), &snap)`

File: client/alloc_runner.go [hunk 4]
```diff
@@ -118,9 +118,28 @@ func (r *AllocRunner) RestoreState() error {
 	return mErr.ErrorOrNil()
 }
 
-// SaveState is used to snapshot our state
+// SaveState is used to snapshot the state of the alloc runner
+// if the fullSync is marked as false only the state of the Alloc Runner
+// is snapshotted. If fullSync is marked as true, we snapshot
+// all the Task Runners associated with the Alloc
 func (r *AllocRunner) SaveState() error {
+	r.saveAllocRunnerState()
```
Comment (dadgar) [inline client/alloc_runner.go]: Put the err in the multi error

**Comment by dadgar** [comment]
LGTM
